File: backend/app/schemas/accreditation_readiness.py
"""Schemas for the Accreditation Readiness Audit endpoints (Stage 13).

Mirrors ``schemas/outcome_alignment.py`` (Stage 12) structurally -- the same
two-component score breakdown shape -- plus the sub-agent readiness
breakdown, findings summary, gaps, and recommendations specific to
accreditation readiness.
"""

from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.models.enums import (
    AgentType,
    AuditRunStatus,
    AuditStatus,
    FileCategory,
    FindingSeverity,
    FindingType,
    ReadinessRiskLevel,
)


# ---------------------------------------------------------------------------
# Finding
# ---------------------------------------------------------------------------


class ReadinessFindingRead(BaseModel):
    """An individual finding from an accreditation readiness run."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    audit_run_id: uuid.UUID
    finding_type: FindingType
    severity: FindingSeverity
    document_category: FileCategory | None = None
    file_id: uuid.UUID | None = None
    title: str
    description: str
    recommendation: str
    is_resolved: bool
    resolved_note: str | None = None
    created_at: datetime


class FindingResolveRequest(BaseModel):
    note: str | None = Field(default=None, max_length=1000)


# ---------------------------------------------------------------------------
# Sub-agent readiness breakdown (items 1-6)
# ---------------------------------------------------------------------------


class SubAgentReadinessRead(BaseModel):
    """Readiness breakdown for one of the six prior compliance agents."""

    group_id: str
    label: str
    agent_type: AgentType
    weight: int
    has_run: bool
    overall_score: float | None = Field(
        default=None,
        description="Latest overall score (0-100) for this agent, or null if not yet run.",
    )
    threshold: float = Field(description="Minimum acceptable score for accreditation readiness.")
    passed: bool = Field(description="True if has_run and overall_score >= threshold.")


# ---------------------------------------------------------------------------
# Findings summary (items 7/8)
# ---------------------------------------------------------------------------


class FindingsSummaryRead(BaseModel):
    """Aggregate counts of findings carried over from prior compliance audits."""

    total: int
    unresolved: int
    critical_total: int
    critical_unresolved: int
    high_unresolved: int


# ---------------------------------------------------------------------------
# Full accreditation readiness report
# ---------------------------------------------------------------------------


class AccreditationReadinessReport(BaseModel):
    """Full structured report for a completed accreditation readiness run.

    Returned by GET /accreditation-readiness-audits/{run_id}/report.
    """

    run_id: uuid.UUID
    module_id: uuid.UUID
    module_code: str
    module_name: str
    academic_year: str
    scope_type: str = Field(description="Entity scope assessed, e.g. 'module'.")

    presence_score: float = Field(description="Weighted sub-agent readiness checklist score (0-100).")
    quality_score: float = Field(description="Evidence-pack completeness + findings-resolution score (0-100).")
    overall_score: float = Field(description="Combined score: 60% presence + 40% quality (0-100).")
    audit_status: AuditStatus
    risk_level: ReadinessRiskLevel = Field(
        description="Accreditation readiness risk level derived from overall score and unresolved findings."
    )

    total_presence_weight: int
    achieved_presence_weight: float
    total_quality_weight: float
    achieved_quality_weight: float

    evidence_completeness_percentage: float = Field(
        description="Live accreditation evidence-pack completeness percentage (Stage 11 reuse).",
        ge=0.0,
        le=100.0,
    )

    sub_agent_readiness: list[SubAgentReadinessRead]
    findings_summary: FindingsSummaryRead

    gaps: list[str] = Field(description="Human-readable accreditation gaps identified for this scope.")
    recommendations: list[str] = Field(description="Deduplicated recommendations to address the gaps.")

    findings: list[ReadinessFindingRead]
    finding_counts: dict[str, int] = Field(
        description="Count of findings per severity: critical, high, medium, low, info."
    )

    summary: str
    generated_at: datetime


# ---------------------------------------------------------------------------
# Audit run schemas
# ---------------------------------------------------------------------------


class AccreditationReadinessRunBrief(BaseModel):
    """Lightweight run info for history lists."""

    model_config = ConfigDict(from_attributes=True, populate_by_name=True)

    id: uuid.UUID
    module_id: uuid.UUID
    agent_type: AgentType
    run_status: AuditRunStatus
    audit_status: AuditStatus | None = None
    overall_score: float | None = Field(default=None, alias="compliance_score")
    documents_present: int | None = None
    documents_missing: int | None = None
    triggered_by_id: uuid.UUID | None = None
    started_at: datetime | None = None
    completed_at: datetime | None = None
    created_at: datetime


class AccreditationReadinessRunRead(BaseModel):
    """Full audit run with findings list (no readiness breakdown; use /report for that)."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    module_id: uuid.UUID
    institution_id: uuid.UUID
    agent_type: AgentType
    run_status: AuditRunStatus
    audit_status: AuditStatus | None = None
    compliance_score: float | None = None
    documents_present: int | None = None
    documents_missing: int | None = None
    total_required: int | None = None
    summary: str | None = None
    error_message: str | None = None
    triggered_by_id: uuid.UUID | None = None
    started_at: datetime | None = None
    completed_at: datetime | None = None
    created_at: datetime
    findings: list[ReadinessFindingRead] = []


# ---------------------------------------------------------------------------
# Trigger response
# ---------------------------------------------------------------------------


class AccreditationReadinessTriggerResponse(BaseModel):
    """Returned immediately when an accreditation readiness audit is triggered."""

    run_id: uuid.UUID
    module_id: uuid.UUID
    status: AuditRunStatus
    message: str

File: frontend/src/lib/api/accreditationReadiness.ts
import type { AuditRunStatus, AuditStatus, FindingSeverity, FindingType, FileCategory, AgentType } from "@/types/enums";

// ---------------------------------------------------------------------------
// Types (mirrors backend schemas/accreditation_readiness.py)
// ---------------------------------------------------------------------------

export interface ReadinessFindingRead {
  id: string;
  audit_run_id: string;
  finding_type: FindingType;
  severity: FindingSeverity;
  document_category: FileCategory | null;
  file_id: string | null;
  title: string;
  description: string;
  recommendation: string;
  is_resolved: boolean;
  resolved_note: string | null;
  created_at: string;
}

export interface SubAgentReadinessRead {
  group_id: string;
  label: string;
  agent_type: AgentType;
  weight: number;
  has_run: boolean;
  overall_score: number | null;
  threshold: number;
  passed: boolean;
}

export interface FindingsSummaryRead {
  total: number;
  unresolved: number;
  critical_total: number;
  critical_unresolved: number;
  high_unresolved: number;
}

export interface AccreditationReadinessReport {
  run_id: string;
  module_id: string;
  module_code: string;
  module_name: string;
  academic_year: string;
  scope_type: string;
  presence_score: number;
  quality_score: number;
  overall_score: number;
  audit_status: AuditStatus;
  risk_level: "low" | "medium" | "high" | "critical";
  total_presence_weight: number;
  achieved_presence_weight: number;
  total_quality_weight: number;
  achieved_quality_weight: number;
  evidence_completeness_percentage: number;
  sub_agent_readiness: SubAgentReadinessRead[];
  findings_summary: FindingsSummaryRead;
  gaps: string[];
  recommendations: string[];
  findings: ReadinessFindingRead[];
  finding_counts: Record<string, number>;
  summary: string;
  generated_at: string;
}

export interface AccreditationReadinessRunBrief {
  id: string;
  module_id: string;
  agent_type: AgentType;
  run_status: AuditRunStatus;
  audit_status: AuditStatus | null;
  overall_score: number | null;
  documents_present: number | null;
  documents_missing: number | null;
  triggered_by_id: string | null;
  started_at: string | null;
  completed_at: string | null;
  created_at: string;
}

export interface AccreditationReadinessRunRead {
  id: string;
  module_id: string;
  institution_id: string;
  agent_type: AgentType;
  run_status: AuditRunStatus;
  audit_status: AuditStatus | null;
  compliance_score: number | null;
  documents_present: number | null;
  documents_missing: number | null;
  total_required: number | null;
  summary: string | null;
  error_message: string | null;
  triggered_by_id: string | null;
  started_at: string | null;
  completed_at: string | null;
  created_at: string;
  findings: ReadinessFindingRead[];
}

export interface AccreditationReadinessTriggerResponse {
  run_id: string;
  module_id: string;
  status: AuditRunStatus;
  message: string;
}

// ---------------------------------------------------------------------------
// API client
// ---------------------------------------------------------------------------

async function apiFetch<T>(path: string, init?: RequestInit): Promise<T> {
  const res = await fetch(`/api/proxy/${path}`, init);
  if (!res.ok) {
    const body = await res.text();
    throw new Error(body || `HTTP ${res.status}`);
  }
  return res.json() as Promise<T>;
}

const BASE = "accreditation-readiness-audits";

export async function triggerAccreditationReadiness(
  moduleId: string,
): Promise<AccreditationReadinessTriggerResponse> {
  return apiFetch<AccreditationReadinessTriggerResponse>(
    `${BASE}/modules/${moduleId}/trigger`,
    { method: "POST" },
  );
}

export async function getLatestAccreditationReadiness(
  moduleId: string,
): Promise<AccreditationReadinessRunRead> {
  return apiFetch<AccreditationReadinessRunRead>(
    `${BASE}/modules/${moduleId}/latest`,
  );
}

export async function getAccreditationReadinessHistory(
  moduleId: string,
  params: { skip?: number; limit?: number } = {},
): Promise<AccreditationReadinessRunBrief[]> {
  const qs = new URLSearchParams();
  if (params.skip != null) qs.set("skip", String(params.skip));
  if (params.limit != null) qs.set("limit", String(params.limit));
  const query = qs.toString();
  return apiFetch<AccreditationReadinessRunBrief[]>(
    `${BASE}/modules/${moduleId}/history${query ? `?${query}` : ""}`,
  );
}

export async function getAccreditationReadinessRun(
  runId: string,
): Promise<AccreditationReadinessRunRead> {
  return apiFetch<AccreditationReadinessRunRead>(`${BASE}/${runId}`);
}

export async function getAccreditationReadinessReport(
  runId: string,
): Promise<AccreditationReadinessReport> {
  return apiFetch<AccreditationReadinessReport>(`${BASE}/${runId}/report`);
}
